src/libAtomVM/nifs.c

```diff
@@ -2031,23 +2069,58 @@ static term nif_erlang_atom_to_binary_2(Context *ctx, int argc, term argv[])
     term atom_term = argv[0];
     VALIDATE_VALUE(atom_term, term_is_atom);
 
-    if (UNLIKELY(argv[1] != LATIN1_ATOM)) {
-        RAISE_ERROR(BADARG_ATOM);
-    }
+    term encoding = argv[1];
 
     GlobalContext *glb = ctx->global;
 
     int atom_index = term_to_atom_index(atom_term);
     size_t atom_len;
     atom_ref_t atom_ref = atom_table_get_atom_ptr_and_len(glb->atom_table, atom_index, &atom_len);
 
+    bool encode_to_latin1 = false;
+    if (encoding == LATIN1_ATOM) {
+        if (UNLIKELY(!atom_table_is_atom_ref_ascii(glb->atom_table, atom_ref))) {
+            encode_to_latin1 = true;
+        }
+    } else if (UNLIKELY(encoding != UTF8_ATOM) && (encoding != UNICODE_ATOM)) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
     if (UNLIKELY(memory_ensure_free_opt(ctx, term_binary_heap_size(atom_len), MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
         RAISE_ERROR(OUT_OF_MEMORY_ATOM);
     }
 
-    term binary = term_create_uninitialized_binary(atom_len, &ctx->heap, glb);
-    atom_table_write_bytes(glb->atom_table, atom_ref, atom_len, (char *) term_binary_data(binary));
-    return binary;
+    if (!encode_to_latin1) {
+        term binary = term_create_uninitialized_binary(atom_len, &ctx->heap, glb);
+        atom_table_write_bytes(
+            glb->atom_table, atom_ref, atom_len, (char *) term_binary_data(binary));
+        return binary;
+    } else {
+        char *utf8_tmp_buf = malloc(atom_len);
+        if (IS_NULL_PTR(utf8_tmp_buf)) {
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+        atom_table_write_bytes(glb->atom_table, atom_ref, atom_len, utf8_tmp_buf);
+        size_t encoded_len = atom_len / 2;
+        term binary = term_create_uninitialized_binary(encoded_len, &ctx->heap, glb);
+        char *binary_data = (char *) term_binary_data(binary);
+        size_t in_pos = 0;
+        for (size_t i = 0; i < encoded_len; i++) {
+            size_t codepoint_size;
+            uint32_t codepoint;
+            if (UNLIKELY(bitstring_utf8_decode(
+                             (uint8_t *) &utf8_tmp_buf[i * 2], 2, &codepoint, &codepoint_size)
```

Comment:
Regarding `i * 2`, we need two cursors here, one for the UTF-8 string and one for the ISO-8859-1 string, the first is not necessary the double of the first.

Reply:
no more `i * 2` and similar assumptions.